### ISSUE TITLE
Update limits file to bump allowed number of file open

### ISSUE THREAD

**Issue by arlake228** [perfsonar/toolkit#293]
I've noticed as of late that even on my CentOS 7 node I am hitting the "number of open files"(nofiles) limit on my busy hosts. Looks to be legitimate and not any processes out of control. We had this problem before on CentOS 6 with number of processes (nprocs) and added a ulimits file to correct it. It was not needed on CentOS 7 because the limits were higher by default. Should be easy enough to update that file with nofiles as well and apply on CentOS 7. The default nofiles limit is 1024, we can bump it to 4096 to match nprocs safely.

**Comment by mfeit-internet2** [comment]
Can confirm separately; I was seeing a lot of this on hosts at IU.

**Comment by tonin** [comment]
After investigation by @vvidic it looks like this needs to be changed on Debian and Ubuntu too as the defaults limits are:

```
Centos 7:
open files                      (-n) 1024
max user processes              (-u) 4096

Debian jessie:
open files                      (-n) 65536
max user processes              (-u) 15797

Debian stretch:
open files                      (-n) 1024
max user processes              (-u) 15736

Ubuntu trusty:
open files                      (-n) 1024
max user processes              (-u) 15726

Ubuntu xenial:
open files                      (-n) 1024
max user processes              (-u) 15663
```

And the current `.deb` packaging doesn't make use of the perfsonar_ulimit.conf file.

@arlake228 I see this file is part of toolkit, but it is activated only in the bundles package. Is that right?

**Comment by arlake228** [comment]
Correct. The file lives in the perfsonar-toolkit-install package but the symlink in  /etc/security/limits.d to the file is created by perfsonar-testpoint. This is consistent with how we handle some other files like the pscheduler limits.

**Comment by vvidic** [comment]
On Tue, Feb 06, 2018 at 01:58:50AM -0800, Antoine Delvaux wrote:
> After investigation by @vvidic it looks like this needs to be changed on Debian and Ubuntu too as the defaults limits are:
> 
> ```
> Centos 7:
> open files                      (-n) 1024
> max user processes              (-u) 4096
> 
> Debian jessie:
> open files                      (-n) 65536
> max user processes              (-u) 15797
> 
> Debian stretch:
> open files                      (-n) 1024
> max user processes              (-u) 15736
> 
> Ubuntu trusty:
> open files                      (-n) 1024
> max user processes              (-u) 15726
> 
> Ubuntu xenial:
> open files                      (-n) 1024
> max user processes              (-u) 15663
> ```
> 
> And the current `.deb` packaging doesn't make use of the perfsonar_ulimit.conf file.

It seems this config will not work on Debian/Ubuntu because:

"Note that pam_limits is not used in /etc/pam.d/common-session and
/etc/pam.d/common-session-noninteractive, so it won't be active for
daemons (it is in PAM configuration for atd, cron, login, sshd, su,
sudo, ...)."
https://wiki.debian.org/Limits

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by vvidic** [comment]
On Tue, Feb 06, 2018 at 01:58:50AM -0800, Antoine Delvaux wrote:
> After investigation by @vvidic it looks like this needs to be changed on Debian and Ubuntu too as the defaults limits are:
> 
> ```
> Centos 7:
> open files                      (-n) 1024
> max user processes              (-u) 4096
> 
> Debian jessie:
> open files                      (-n) 65536
> max user processes              (-u) 15797
> 
> Debian stretch:
> open files                      (-n) 1024
> max user processes              (-u) 15736
> 
> Ubuntu trusty:
> open files                      (-n) 1024
> max user processes              (-u) 15726
> 
> Ubuntu xenial:
> open files                      (-n) 1024
> max user processes              (-u) 15663
> ```
> 
> And the current `.deb` packaging doesn't make use of the perfsonar_ulimit.conf file.

Also limits.conf might be ignored by systemd?

https://support.hpe.com/hpsc/doc/public/display?docId=mmr_kc-0128457

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by tonin** [comment]
We discussed this in the dev call today and this setup is working on the ESnet hosts where Andy was seeing this issue.  @vvidic can you try to reproduce the problem under Debian and see what configuration change needs to be done to solve any issue?

**Comment by vvidic** [comment]
On Wed, Feb 07, 2018 at 04:57:04PM +0000, Antoine Delvaux wrote:
> We discussed this in the dev call today and this setup is working on
> the ESnet hosts where Andy was seeing this issue.  @vvidic can you try
> to reproduce the problem under Debian and see what configuration
> change needs to be done to solve any issue?

I could not get limits.conf to work on Debian/Ubuntu for daemons,
because they don't got through PAM when switching users from root.
I can test a bit more, but it is possible that RHEL has some patch
to enable this.

The solution seems to be setting limits in systemd service files
and init scripts, but I would need to know which services need
this fix?

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by tonin** [comment]
After discussing with @mfeit-internet2 it seems we can limit that to the pscheduler services.

**Comment by vvidic** [comment]
On Wed, Feb 07, 2018 at 09:36:36AM -0800, Antoine Delvaux wrote:
> After discussing with @mfeit-internet2 it seems we can limit that to
> the pscheduler services.

Ok, I will prepare a patch for pscheduler-* services.

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by arlake228** [comment]
Also need to do owamp-server, definitely seen the owamp user hit the limit and that user in particular was giving Shawn McKee some trouble a couple weeks ago. It'd be nice if we could do apache as well, but since we don't control that  we could see how things hold-up without it.

Ultimately, perhaps we should do this in systemd as well for CentOS so we are consistent. In theory it should still work. It's not clear to me just reading which way is supposed to work on CentOS (I am finding claims in both directions). My own observations are that ulimits still works on the ESnet CentOS 7 hosts, but if it also works in systemd that may be cleaner.

**Comment by vvidic** [comment]
On Wed, Feb 07, 2018 at 06:18:12PM +0000, Andrew R. Lake wrote:
> Also need to do owamp-server, definitely seen the owamp user hit the
> limit and that user in particular was giving Shawn McKee some trouble
> a couple weeks ago. It'd be nice if we could do apache as well, but
> since we don't control that  we could see how things hold-up without
> it.

It should be possible to modify external services using drop-in files
as described in systemd.unit(5):

  Alternatively, the administrator could create a drop-in file
  /etc/systemd/system/httpd.service.d/local.conf with the following
  contents:

           [Unit]
           After=memcached.service
           Requires=memcached.service
           # Reset all assertions and then re-add the condition we want
           AssertPathExists=
           AssertPathExists=/srv/www

> Ultimately, perhaps we should do this in systemd as well for CentOS so
> we are consistent. In theory it should still work. It's not clear to
> me just reading which way is supposed to work on CentOS (I am finding
> claims in both directions). My own observations are that ulimits still
> works on the ESnet CentOS 7 hosts, but if it also works in systemd
> that may be cleaner. 

There might be some explanations here if someone has access :)

  https://access.redhat.com/solutions/1257953

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by vvidic** [comment]
On Wed, Feb 07, 2018 at 06:18:12PM +0000, Andrew R. Lake wrote:
> Also need to do owamp-server, definitely seen the owamp user hit the
> limit and that user in particular was giving Shawn McKee some trouble
> a couple weeks ago. It'd be nice if we could do apache as well, but
> since we don't control that  we could see how things hold-up without
> it.

Strange thing is I don't have it working on centos7:

[root@centos7 ~]# cat /etc/security/limits.d/perfsonar_ulimit.conf 
# Increases process limit and number of files open for various perfsonar
# tools

pscheduler  soft    nproc     4096
apache      soft    nproc     4096
pscheduler  soft    nofile    4096
apache      soft    nofile    4096
perfsonar   soft    nofile    4096
owamp       soft    nofile    4096

[root@centos7 ~]# systemctl restart owamp-server
[root@centos7 ~]# systemctl status owamp-server
● owamp-server.service - OWAMP server
   Loaded: loaded (/usr/lib/systemd/system/owamp-server.service; enabled; vendor preset: disabled)
   Active: active (running) since Wed 2018-02-07 14:13:08 EST; 3s ago
  Process: 2433 ExecStart=/usr/bin/owampd -c /etc/owamp-server -R /var/run (code=exited, status=0/SUCCESS)
 Main PID: 2434 (owampd)
   CGroup: /system.slice/owamp-server.service
           └─2434 /usr/bin/owampd -c /etc/owamp-server -R /var/run

[root@centos7 ~]# grep files /proc/2434/limits
Max open files            1024                 4096 files     

So owamp soft limit still on 1024 even after reboot?

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by arlake228** [comment]
Did you reboot your host or just restart the service? I usually reboot the host after changing the ulimits.

**Comment by vvidic** [comment]
On Wed, Feb 07, 2018 at 07:25:57PM +0000, Andrew R. Lake wrote:
> Did you reboot your host or just restart the service? I usually reboot
> the host after changing the ulimits.

Yes, I did an OS upgrade now + reboot and still no change for open files
in owampd:

[root@centos7 ~]# cat /proc/887/comm 
owampd

[root@centos7 ~]# cat /proc/887/limits
Limit                     Soft Limit           Hard Limit           Units     
Max cpu time              unlimited            unlimited            seconds   
Max file size             unlimited            unlimited            bytes     
Max data size             unlimited            unlimited            bytes     
Max stack size            8388608              unlimited            bytes     
Max core file size        0                    unlimited            bytes     
Max resident set          unlimited            unlimited            bytes     
Max processes             7280                 7280                 processes 
Max open files            1024                 4096                 files     
Max locked memory         65536                65536                bytes     
Max address space         unlimited            unlimited            bytes     
Max file locks            unlimited            unlimited            locks     
Max pending signals       7280                 7280                 signals   
Max msgqueue size         819200               819200               bytes     
Max nice priority         0                    0                    
Max realtime priority     0                    0                    
Max realtime timeout      unlimited            unlimited            us        

Does the /proc/PID/limits look good on your servers?

-- 
Valentin Vidic
Computer Systems Engineer - Expert
Department of Computer Infrastructure and Services
Croatian Academic and Research Network - CARNET
Josipa Marohnica 5, HR-10000 Zagreb, Croatia
tel: +385 1 6661 714, fax. +385 1 6661 635
gsm: +385 91 2480 919
www.CARNet.hr

**Comment by arlake228** [comment]
So /proc/PID/limits on my system looks the same as yours, but lsof seems to indicate more than 1024 files are open. It's also true that before the update I was hitting the limit a short-time after reboot, and since updating ulimits i have not had any issue on our ESnet hosts. I am officially confused, so let's try the systemd method and see if we get results that make more sense. I merged in everything but the pscheduler change, pending Mark's approval.

**Comment by mfeit-internet2** [comment]
Dan reports continued problems on the I2 PAS machines.  I did some tracing of the running processes and am finding sporadic `EAGAIN` errors on calls to `futex()`.

Futexes lock down an integer's worth of memory where they live; the kernel locks memory in 4K pages.  The default maximum is 64K, so in a worst-case situation where the locked areas are all in different pages, there would be at most 16 before reaching the limit.

I recommended trying 256K or 512K; there's probably no reason why for that user it can't be something much larger or unlimited.
